File: it-agents/skills/code-architect/references/code-examples/cpp/Bridge.cpp
// Bridge — split abstraction and implementation into independent hierarchies.
#include <iostream>
#include <memory>
#include <string>

// Implementation hierarchy.
class Backend {
public:
    virtual ~Backend() = default;
    virtual std::string emit(const std::string& msg) const = 0;
};

class ConsoleBackend : public Backend {
public:
    std::string emit(const std::string& msg) const override { return "[console] " + msg; }
};

class FileBackend : public Backend {
public:
    std::string emit(const std::string& msg) const override { return "[file] " + msg; }
};

// Abstraction holds a reference to an implementation.
class Reporter {
protected:
    std::shared_ptr<Backend> backend;
public:
    explicit Reporter(std::shared_ptr<Backend> b) : backend(std::move(b)) {}
    virtual ~Reporter() = default;
    virtual void report(const std::string& text) const = 0;
};

// Refined abstraction.
class PlainReporter : public Reporter {
public:
    using Reporter::Reporter;
    void report(const std::string& text) const override {
        std::cout << backend->emit(text) << "\n";
    }
};

class UrgentReporter : public Reporter {
public:
    using Reporter::Reporter;
    void report(const std::string& text) const override {
        std::cout << backend->emit("!! " + text + " !!") << "\n";
    }
};

int main() {
    auto console = std::make_shared<ConsoleBackend>();
    auto file = std::make_shared<FileBackend>();

    PlainReporter{console}.report("startup ok");
    UrgentReporter{file}.report("disk low");
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/ruby/bridge.rb
# frozen_string_literal: true

# Bridge — split abstraction and implementation into independent hierarchies.

# Implementation hierarchy: how output is physically emitted.
class ConsoleSink
  def emit(line) = "console> #{line}"
end

class FileSink
  def initialize(path) = @path = path
  def emit(line) = "#{@path}: #{line}"
end

# Abstraction hierarchy: what is logged, holding a reference to an implementation.
class Logger
  def initialize(sink) = @sink = sink

  def log(message) = @sink.emit(message)
end

# Refined abstraction extends behavior without touching sinks.
class TimestampLogger < Logger
  def log(message) = super("[#{stamp}] #{message}")

  private

  def stamp = "T+0"
end

if __FILE__ == $PROGRAM_NAME
  puts Logger.new(ConsoleSink.new).log("plain")
  puts TimestampLogger.new(ConsoleSink.new).log("with time")
  puts TimestampLogger.new(FileSink.new("/var/log/app")).log("to file")
end
